core/src/main/java/org/apache/iceberg/InternalData.java

```diff
@@ -0,0 +1,159 @@
+    READ_BUILDERS.put(format, readBuilder);
+  }
+
+  public static WriteBuilder write(FileFormat format, OutputFile file) {
+    Function<OutputFile, WriteBuilder> writeBuilder = WRITE_BUILDERS.get(format);
+    if (writeBuilder != null) {
+      return writeBuilder.apply(file);
+    }
+
+    throw new UnsupportedOperationException(
+        "Cannot write using unregistered internal data format: " + format);
+  }
+
+  public static ReadBuilder read(FileFormat format, InputFile file) {
+    Function<InputFile, ReadBuilder> readBuilder = READ_BUILDERS.get(format);
+    if (readBuilder != null) {
+      return readBuilder.apply(file);
+    }
+
+    throw new UnsupportedOperationException(
+        "Cannot read using unregistered internal data format: " + format);
+  }
+
+  public interface WriteBuilder {
+    /** Set the file schema. */
+    WriteBuilder schema(Schema schema);
+
+    /** Set the file schema's root name. */
+    WriteBuilder named(String name);
+
+    /**
+     * Set a writer configuration property.
+     *
+     * <p>Write configuration affects writer behavior. To add file metadata properties, use {@link
+     * #meta(String, String)}.
+     *
+     * @param property a writer config property name
+     * @param value config value
+     * @return this for method chaining
+     */
+    WriteBuilder set(String property, String value);
+
+    /**
+     * Set a file metadata property.
+     *
+     * <p>Metadata properties are written into file metadata. To alter a writer configuration
+     * property, use {@link #set(String, String)}.
+     *
+     * @param property a file metadata property name
+     * @param value config value
+     * @return this for method chaining
+     */
+    WriteBuilder meta(String property, String value);
+
+    /**
+     * Set a file metadata properties from a Map.
+     *
+     * <p>Metadata properties are written into file metadata. To alter a writer configuration
+     * property, use {@link #set(String, String)}.
+     *
+     * @param properties a map of file metadata properties
+     * @return this for method chaining
+     */
+    default WriteBuilder meta(Map<String, String> properties) {
+      properties.forEach(this::meta);
+      return this;
+    }
+
+    /** Overwrite the file if it already exists. */
+    WriteBuilder overwrite();
+
+    /** Build the configured {@link FileAppender}. */
+    <D> FileAppender<D> build() throws IOException;
+  }
+
+  public interface ReadBuilder {
+    /** Set the projection schema. */
+    ReadBuilder project(Schema projectedSchema);
+
+    /** Read only the split that is {@code length} bytes starting at {@code start}. */
+    ReadBuilder split(long start, long length);
+
+    /** Reuse container classes, like structs, lists, and maps. */
+    ReadBuilder reuseContainers();
+
+    /** Set a custom class for in-memory objects at the schema root. */
+    ReadBuilder setRootType(Class<? extends StructLike> rootClass);
```

Comment:
I'll probably get to this later in the PR but i'm interested in why we need this and setCustomType

Reply:
Ok I see how it's used below, I'm wondering if instead of needing this, could we just automatically set these readers based on the root type?  Ie

setRootType(ManifestEntry) --- Automatically sets field types based on Manifest entry?

Or do we have a plan for using this in a more custom manner in the future?

Reply:
The problem this is solving is that we don't have an assigned ID for the root type. We could use a sentinel value like `-1`, but that could technically collide. I just don't want to rely on `setCustomType(ROOT_FIELD_ID, SomeObject.class)`.